examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js

```diff
@@ -0,0 +1,161 @@
+import WebGPUUniformsGroup from './WebGPUUniformsGroup.js';
+import { FloatUniform, Vector3Uniform } from './WebGPUUniform.js';
+
+import NodeSlot from '../nodes/core/NodeSlot.js';
+import NodeBuilder from '../nodes/core/NodeBuilder.js';
+
+class WebGPUNodeUniformsGroup extends WebGPUUniformsGroup {
+
+	constructor( shaderStage ) {
+
+		super( 'nodeUniforms' );
+
+		let shaderStageVisibility;
+
+		if ( shaderStage === 'vertex' ) shaderStageVisibility = GPUShaderStage.VERTEX;
+		else if ( shaderStage === 'fragment' ) shaderStageVisibility = GPUShaderStage.FRAGMENT;
+
+		this.setVisibility( shaderStageVisibility );
+
+		//this.setOnBeforeUpdate( this._onBeforeUpdate );
+
+	}
+	/*
+	_onBeforeUpdate( object, camera ) {
+
+		const material = object.material;
+
+	}
+	*/
+
+}
+
+class WebGPUNodeBuilder extends NodeBuilder {
+
+	constructor( material, renderer ) {
+
+		super( material, renderer );
+
+		this.uniformsGroup = {};
+
+		this._parseMaterial();
+
+	}
+
+	_parseMaterial() {
+
+		const material = this.material;
+
+		if ( material.isMeshBasicMaterial || material.isPointsMaterial ) {
+
+			if ( material.colorNode !== undefined ) {
+
+				this.addSlot( 'fragment', new NodeSlot( material.colorNode, 'COLOR', 'vec3' ) );
+
+			}
+
+			if ( material.opacityNode !== undefined ) {
+
+				this.addSlot( 'fragment', new NodeSlot( material.opacityNode, 'OPACITY', 'float' ) );
+
+			}
+
+		}
+
+	}
+
+	getUniformNSName( nodeUniform ) {
+
+		return `nodeUniforms.${nodeUniform.name}`;
+
+	}
+
+	getBindings() {
+
+		const bindings = [];
+
+		const uniformsVertexGroup = this.uniformsGroup[ 'vertex' ];
+		const uniformsFragmentGroup = this.uniformsGroup[ 'fragment' ];
+
+		if ( uniformsVertexGroup ) bindings.push( uniformsVertexGroup );
+		if ( uniformsFragmentGroup ) bindings.push( uniformsFragmentGroup );
+
+		return bindings;
+
+	}
+
+	getUniformFromNode( node, shaderStage, type ) {
+
+		const uniformNode = super.getUniformFromNode( node, shaderStage, type );
+		const nodeData = this.getDataFromNode( node, shaderStage );
+
+		if ( nodeData.uniformGroup === undefined ) {
```

Comment:
Is this check correct? If I'm right, `nodeData.uniformGroup` seems always `undefined` because `.uniformGroup` doesn't seem to be assigned anywhere (`$ find ./examples/jsm/renderers/ -print0 | xargs -0 grep uniformGroup` matches only this line).

Reply:
Fixed! Thanks.